### DISCUSSION TITLE
Feature to Keep Input on ESC

### DISCUSSION THREAD

**Discussion by laktak** [andmarti1424/sc-im#709] [General]
I'd like to request a config option that makes ESC keep the input when going from edit mode to normal mode.

At the moment ESC will undo any edits (with no way to get them back AFAIK). For me this is very error prone because I often forget to ESC-ENTER from input mode instead of just ESC-ESC which is more of what I am used to from vim. Also see https://github.com/andmarti1424/sc-im/discussions/617

I don't think this can be done at the moment as there are not mappings for edit mode.

Would you accept a PR for this feature? What should be the name of the config option? (keepesc?)

**Comment by andmarti1424** [comment]
Hello. Yes. PR are welcome. I will also think about this.

**Comment by andmarti1424** [comment]
@laktak I decided to modify this so ESC key also confirm changes. Please feel free to update to latest commit on dev branch and try it.

**Comment by andmarti1424** [comment]
@laktak could you test it? Thanks.

**Comment by laktak** [comment]
Yes it works great, thank you!

Could you make the same change when entering a new text or number? 

When I enter something with `<` or `=` and then press `ESC` the value will be discarded. IMO it should go to edit mode. What do you think?

**Comment by andmarti1424** [comment]
I dont think so. gnumeric and libreoffice also work like sc-im in those situations.

**Comment by laktak** [comment]
Interesting. What is the reason for this behavior?

So when entering text I first do `<` `....` `enter` then `E` so I am in edit mode?

**Comment by andmarti1424** [comment]
Yes, indeed.
sc-im starts in NORMAL mode.
with `<` you enter INSERT_MODE. you can enter content and move cursor with arrow keys and delete content with delete key.
If you press `ESC` you go back to NORMAL_MODE.
If then you press `E` you enter EDIT_MODE so you can edit content as you normally do in vim.

**Comment by laktak** [comment]
I understand how it works but not why. What is the advantage?

Wouldn't it be easier to use the app if it always behaves the same way? So when pressing ESC in insert mode you always get to edit mode (instead of losing the text you entered as it is at the moment)?

Would you allow `E` to go to insert/edit mode if there is no text instead of showing "No string value to edit"?

**Comment by andmarti1424** [comment]
The later, yes.. I think that could be modified.

I treat NORMAL_MODE as the default mode. That means where you most of the time are. (I think vim also has this concept). So when you 're in INSERT_MODE I believe ESC should get back to your home. That is NORMAL_MODE..

**Comment by laktak** [comment]
I can see your argument except for me EDIT_MODE maps to vim's normal mode and NORMAL_MODE is more like a mode above that. That is why I feel that going from insert to sc-im's normal mode is skipping a  step.

Please take a look at my PR that allows the user to select the desired behavior.

**Comment by andmarti1424** [comment]
I move this to a discussion. What other people think about this..?
What should ESC key do in edit mode, cancel the changes or confirm them?

**Reply by laktak** [reply]
I've been using the patch since last year and it works great IMO.

With ESC always exiting a mode and no longer canceling my input I never lost any text I entered. And for accidental changes there is always `u`/undo.
